handle magics

densnow opened the pull request:
Magics no longer cause a cell to be skipped when passing `allowed` Jupyter notebooks as long as the user has `ipython`  installed. Cells with syntax errors will report a "syntax error" for the corresponding cell number on line number 1.

There is a couple more little changes: I have "tidied up" the process of printing warnings to the user after the program has run, and updated the README. The messages and README might need a few small tweaks for personal taste here and there but I have covered the gist of it.

mwermelinger commented:
No need to squash commits. I do it when merging the PR.

densnow commented:
Ah, OK I will stop doing that, still learning the ropes and way things are done here on Github!

mwermelinger commented on allowed.py:
```diff
@@ -524,32 +549,43 @@ def check_file(filename: str, constructs: tuple, check_method_calls: bool) -> No
-                pass
+            except SyntaxError:
+                line_cell_map[syntax_error_num] = (cell_num, 1)
+                syntax_error_num -= 1
```
You're using the same dictionary for two different purposes with two different keys: line numbers and error numbers. That will lead to hard-to-debug errors. Separate the purposes, e.g. use a simple list to remember which cells have errors, and add arguments to functions if necessary, but don't shoehorn multiple purposes into the same data structure.

densnow replied:
Yes I can see your point and I will change this. This is probably one of those habits I have  from solving to many leetcode problems and other puzzles where thoughts about longevity of the code and maintainability are often non existent.

densnow replied:
This has been changed now so that `read_jupyter_notebooks` returns a tuple of 3 objects. This got me thinking about when (or if) there is a time to start thinking about bundling the information about Jupyter notebooks into some kind of class. `read_jupyter_notebooks` could be replaced with the constructor and `get_line` could be encapsulated to. Moreover, there could be a time in the future when we want to allow or disallow individual magics and report the errors back as magics (as apposed to get_ipython() calls) and it might make more sense to have these as methods of some kind of object.

At the same time I appreciate that this is just a small script to check if certain constructs are allowed so this might be complete overkill. what do you think? is this something that could be done in the future?

mwermelinger commented on allowed.py:
```diff
@@ -622,5 +659,5 @@ def read_jupyter_notebook(file_contents: str) -> tuple:
-    if reminder:
-        print(reminder)
+    for warning in warnings.values():
+        print(warning)
```
This could be `print("WARNING:", warning)` to make strings shorter in the code.

densnow replied:
Good idea, this is one of those little tricks of the trade to remember!

densnow commented on README.md:
```diff
@@ -115,29 +115,16 @@ anywhere after `allowed.py` and in either form.
+magics such as `%timeit` and `%run` will trigger a syntax error in the
+absence of a `ipython` installation. Conversely, if `ipython` is installed,
+the 'magics' will be deemed valid Python, allowing those cells to be checked
+(other syntax errors not withstanding).
```
Well I would (cheekily) argue that they are still valid Python when transformed no matter what is allowed :smile:, but I absolutely get your point and will change this to mention the function calls and attributes.

densnow commented on allowed.py:
```diff
@@ -7,6 +7,9 @@
 import sys
 
+# warnings will be printed after the progam has run.
+warnings = {}
```
This whole dictionary thing was a messy fix to an issue I discovered after submitting the PR. I realised that the message about cells would be printed for checks that were only done on .py files. Needless to say it had not been thought through properly and the dict was used so I could remove a warning, but as you have seen there is an issue with that too. 


I guess I should have just mentioned the first issue in a comment in this thread and not tried to quickly fix it :smile:

densnow commented on allowed.py:
```diff
@@ -492,6 +515,8 @@ def check_file(filename: str, constructs: tuple, check_method_calls: bool) -> No
                 source, line_cell_map = read_jupyter_notebook(file.read())
             else:
+                if "ipython" in warnings:
+                    del warnings["ipython"]
```
Yes you are correct this is an error. your suggestion is the fix I was looking for, but never quite found.

densnow commented:
All changes have been made as per comments. The README may still need a bit of a tweak but I think I have addressed most of the suggestions and comments.
